darts/models/filtering/kalman_filter.py

```diff
@@ -4,150 +4,193 @@
+            self.dim_u = covariates.width
+            inputs = covariates.pd_dataframe()
+            inputs.columns = [f'u_{i}' for i in inputs.columns]
+            input_columns = list(inputs.columns)
+            measurements = pd.concat([outputs, inputs], axis=1)
+        else:
+            measurements = outputs
+            input_columns = None
+
+        if num_block_rows is None:
+            num_block_rows = max(10, self.dim_x)
+        nfoursid = NFourSID(measurements,
+                            output_columns=list(outputs.columns),
+                            input_columns=input_columns,
+                            num_block_rows=num_block_rows)
+        nfoursid.subspace_identification()
+        state_space_identified, covariance_matrix = nfoursid.system_identification(
+            rank=self.dim_x
+        )
+
+        self.kf = Kalman(state_space_identified, covariance_matrix)
+
+
     def filter(self,
                series: TimeSeries,
-               num_samples: int = 1):
+               covariates: Optional[TimeSeries] = None,
+               num_samples: int = 1) -> TimeSeries:
         """
         Sequentially applies the Kalman filter on the provided series of observations.
 
         Parameters
         ----------
         series : TimeSeries
-            The series of observations used to infer the state values according to the specified Kalman process.
+            The series of outputs (observations) used to infer the underlying outputs according to the specified Kalman process.
+            This must be a deterministic series (containing one sample).
+        covariates : Optional[TimeSeries]
+            An optional series of inputs (control signal), necessary if the Kalman filter was initialized with covariates.
             This must be a deterministic series (containing one sample).
 
         Returns
         -------
         TimeSeries
-            A stochastic `TimeSeries` of state values, of dimension `dim_x`.
+            A stochastic `TimeSeries` of state values, of the same width as the input series.
         """
+        super().filter(series)
 
-        raise_if_not(series.is_deterministic, 'The input series for the Kalman filter must be '
-                                              'deterministic (observations).')
+        raise_if(self.kf is None, 'The Kalman filter has not been fitted yet. Call `fit()` first '
+                                  'or provide Kalman filter in constructor.')
+                                  
+        raise_if_not(series.width == self.dim_y, 'The provided TimeSeries dimensionality does not match '
+                                                 'the output dimensionality of the Kalman filter.')
 
-        dim_z = series.width
+        if self._expect_covariates:
+            raise_if(covariates is None and self._expect_covariates,
+                    'The Kalman filter was fitted with covariates, but these were not provided.')
```

Comment:
some raisers in this file are under-indented.
```suggestion
                     'The Kalman filter was fitted with covariates, but these were not provided.')
```